M3-2353 Part III

rjstires opened the pull request:
## Description
Refactoring LinodeDetail to source information from Redux.
The ebb and flow of this is using the LinodesDetail.container pipeline to build the data object, then in LinodesDetail combine that data object with Redux thunks, and "provide" it to the lower DOM via ContextAPI.

[You should read this.](https://github.com/linode/manager/blob/ba86042868a5a25d10d1576151ae421fa8116364/src/features/linodes/LinodesDetail/README.md)

## Todo
Currently Linode configs and disks are only retrieved when the component mounts or updates. As users add/update/delete configs, they are not added to that internal state and are not represented accurately in the application.
- [x] Add LinodeDisks to Redux. 
  - [x] dispatch `getAllLinodeDisks` when `disk_create`, `disk_delete`, or `disk_resize` events are received.
- [x] Add LinodeConfigs to Redux.
  - [x] Need to remove all related configs and disks when a Linode is deleted.
  - [ ] Reset disk state (Remove all disks and getAll) when a rebuild finishes.

### Other (Can be tech story tickets for later time)
- [ ] LinodeConfigs.tsx can be updated to use Paginate and OrderBy components, rather than pagey, now.
- [ ] Most/many references to getLinodeConfig(s) can be updated to use the Thunk library, which would help keeping the store up-to-date.
- [ ] Most/many references to getLinodeDisks can be updated to use the Thunk library, which would help keeping the store up-to-date.
- [ ] Selectors `getVolumesForLinode`, `getNotificationsForLinode`, `getTypeById`, `eventsForLinode`, `getLinodeConfigsForLinode`, and `getLinodeDisksForLinode` should be memoized with reselect.
## Type of Change
- Non-breaking change

## Note to Reviewers
Godspeed.

martinmckenna commented on src/features/linodes/LinodesDetail/LinodeSettings/LinodeConfigDrawer.tsx:
```diff
@@ -780,3 +731,63 @@ const isUsingCustomRoot = (value: string) =>
+  })
+);
+
+export default enhanced(LinodeConfigDrawer);
```
note to self: there is no error handling at all in this drawer so we need to make a ticket for it

rjstires commented:
> * if you block requests -> load a detail page (see an error) -> unblock requests -> navigate away and then back, the error is still there.

Fixed @Jskobos. I've also renamed the lifecycle HOCs to be (maybe) a little more obvious about what they do and moved documentation into the files themselves.

martinmckenna commented on src/features/linodes/LinodesDetail/LinodeSettings/LinodeSettingsLabelPanel.tsx:
```diff
@@ -126,17 +108,10 @@ const styled = withStyles(styles);
   styled,
-  linodeContext,
-  withLinodeActions
+  linodeContext
```
How come the whole object is now being passed down instead of just the label and ID?

rjstires replied:
I know I was a little inconsistent here, but my intention is to pass the entire context object and destructure as necessary in the component. This would be the conversion to useLinodeDetailContext very very straightforward.

martinmckenna commented on src/features/linodes/LinodesDetail/LinodesDetailHeader/LinodeControls.tsx:
```diff
@@ -0,0 +1,166 @@
+    }
+  };
+
+  const handleSubmitLabelChange = (label: string) => {
```
this isn't handling non API errors (i.e request blocking)

rjstires replied:
From what I can see, this is existing and should certainly be resolved in another PR.

martinmckenna commented on src/features/linodes/LinodesDetail/LinodesDetailNavigation.tsx:
```diff
@@ -0,0 +1,151 @@
+import { compose } from 'recompose';
+import 'rxjs/add/observable/timer';
+import 'rxjs/add/operator/debounce';
+import 'rxjs/add/operator/filter';
```
unused rxjs imports

rjstires replied:
Good catch! Pushing fix.

martinmckenna commented:
Just got through this. Overall, I think it looks good, but I'd like to come back later with a clearer mind.

Also I wouldn't be remiss if I didn't bring up the fact that clicking around the navigation feels noticeably slower than before. Looks like some onClick handlers are taking nearly 700ms to complete. This is definitely a tradeoff with preloaded data, but I feel like we should have some plan going forward to increase the perceived time.

Maybe we can look into ways to prevent the navigation lag even though the data is still loading.

rjstires commented:
> Just got through this. Overall, I think it looks good, but I'd like to come back later with a clearer mind.
> 
> Also I wouldn't be remiss if I didn't bring up the fact that clicking around the navigation feels noticeably slower than before. Looks like some onClick handlers are taking nearly 700ms to complete. This is definitely a tradeoff with preloaded data, but I feel like we should have some plan going forward to increase the perceived time. In other words, make it look like it's loading faster even though it's not
> 
> One example would be to immediately highlight the newly selected route and THEN get the data from redux - I'm not sure if this is possible but just a suggestion. As it stands now, the navigation highlighting is very laggy

1. You **would** be remiss...
2. Which onClick handles are you referring to? There is a note (above) about using Reselect. It's likely that the context is being recreated unnecessarily.

johnwcallahan commented on src/features/linodes/LinodesDetail/LinodeBackup/LinodeBackup.tsx:
```diff
@@ -759,7 +759,7 @@ const mapStateToProps: MapState<StateProps, {}> = state => ({
 const connected = connect(mapStateToProps);
 
-const linodeContext = withLinode(({ linode }) => ({
+const linodeContext = withLinodeDetailContext(({ linode }) => ({
```
After enabling (or disabling) backups, this component is not updated, i.e. the component does not change from the placeholder to the actual controls.

1) Create a Linode.
2) Wait until the Linode is finished provisioning (you may need to refresh).
3) Click the Backups tab.
4) Click "Enable Backups".
5) **Observe: the button is stuck in the "submitting" state*.*
6) Refresh the page.
7) **Observe: the backup controls/settings are displayed.**

rulikrulit commented:
Upgrade experience is broken in this branch. `click here` is not functional. + minor visual bug with unwanted `(`
<img width="1419" alt="screen shot 2019-02-19 at 10 45 38 am" src="https://user-images.githubusercontent.com/6481831/53028339-ab655780-3434-11e9-8d9d-de7ed7da13f8.png">
